File: java/challenges/lib/src/main/java/challenges/stacksandqueues/PseudoQueue.java
package challenges.stacksandqueues;

import java.util.NoSuchElementException;

public class PseudoQueue {
  public Node top;
  Stack stackOne;
  Stack stackTwo;

  public PseudoQueue(){
    this.top = null;
    stackOne = new Stack();
    stackTwo = new Stack();
  }

  public void pseudoEnqueue(Object val){
      stackOne.push(val);
    }

  public Object pseudoDequeue(){
    Object val;
    if (stackTwo.top == null){
      if (stackOne.top == null){
        throw new NoSuchElementException();
      } else {
        while (stackOne.top != null){
          stackTwo.push(stackOne.pop());
        }val = stackTwo.pop();
      }
    } else {
      val = stackTwo.pop();
    }return val;
  }
  // review: just enqueue everything into stack one, and dequeue in stack 2 only. Stack 1 only pushes into 2
  // when 2 is empty.

}

File: javascript/stackAndQueue/pseudoQueue.js
'use strict';

const { Stack } = require('./stack');

class PseudoQueue {

  constructor()
  {
    this.stackOne = new Stack();
    this.stackTwo = new Stack();
  }

  enqueue(val) {
    while (this.stackTwo.top) {
      this.stackOne.push(this.stackTwo.pop());
    }
    if (!this.stackTwo.top){
      this.stackOne.push(val);
    }
  }

  dequeue() {
    let deQ;
    if (!this.stackTwo.top) {
      if (!this.stackOne.top){
        return 'stacks empty';
      } else {
        while (this.stackOne.top) {
          let temp = this.stackOne.top.next;
          console.log('stack one TOP: ' + this.stackOne.top.val);
          this.stackTwo.push(this.stackOne.pop());
          this.stackOne.top = temp;
          console.log('stack two TOP: ' + this.stackTwo.top.val);
        } deQ = this.stackTwo.pop();
      }
    } else {
      deQ = this.stackTwo.pop();
    } return deQ;
  }
}

module.exports = { PseudoQueue };
